realm/rfc/0001-backlink-query-syntax.md

```diff
@@ -0,0 +1,246 @@
+- Feature Name: Inverse relationship syntax
+- Start Date: 2016-05-30
+- RFC PR: 
+- Version: 4
+
+# Summary
+
+It should be possible to define and query backlinks in both `Realm` and 
+`DynamicRealm`. As a `DynamicRealm` can work without _any_ Realm model 
+classes, it should be possible to query backlinks without them being defined there.
+
+# Motivation
+
+Backlinks are special in Realm. They are not defined explicitly in the schema,
+but are automatically created when defining Links and LinkLists (forward links).
+
+This creates some conflicts with the current query system as it rely on all 
+fields being defined by the schema. This is called *named backlinks*.
+
+This proposal introduces a new syntax for queries so it is possible to query
+across a backlink even though it hasn't been defined by a model class. This is
+called *unnamed backlinks*.
+
+Realm Objective C/Swift shipped with only support for named backlinks. This was mostly
+prompted by restrictions in NSPredicate, which do not have a syntax for unnamed backlinks.
+
+Realm Java does not have this constraint and at the same time the Java Migration/Dynamic API
+is more full-fledged and expected to have the same feature set as the static API due to how the
+query api works.
+
+Adding support for the unnamed variant now will also most likely save us time documenting the
+difference as well as reduce the amount of support we have to do.
+
+Any design should ideally solve all these constraints:
+
+- Should support a list of backlink objects.
+
+- Should support a single backlink object, enforcing the constraint that only 1 object
+  can point to it. This is not supported by Core, but it is a natural extension.
+
+- Should be usable from the Dynamic API.
+
+- Should either support polymorphic objects or not prevent them from being 
+  implemented.
+
+- Ideally it should work for both the pure text-based variant of the query language in the
+  Object Store as well as the current mixed style found in Realm Java.
+  
+
+# Detailed design
+
+**Example model classes:**
+
+```
+public class Person extends RealmObject {
+	private Dog favoriteDog;
+	private RealmList<Dog> dogs;
+}
+
+public class Dog extends RealmObject {
+	private String name
```

Comment:
> I am perceiving that we are to look for `Person` rather than `Dog` when we're to search `realm.where(Dog.class).equalTo("favoriteOwner.name", "John").findAll();`.

I understood it to be a query for `Dog` since it has `where(Dog.class)`. That indicates some confusion on one or both of our sides. 😉